Replication state

sethvincent opened the pull request:
Closes #51.

This pr provides progress tracking of replication between a core and its peers.

Based largely on https://github.com/digidem/hypercore-replication-progress/

gmaclennan commented:
This will need some adjustments if #61 is the way forward.

Also maybe we can brainstorm how this might work for the blob store? I wrote an initial implementation in #62. I think we might need to handle it differently because replication being "complete" is not as simple for the blob store - for other stores our aim is to sync everything, so when the peer and local have everything shared, we consider replication "done". However with the blobs we will only be syncing previews and thumbnails from other peers (not originals) when syncing phones, and in the future will have more complicated selective sync. Also we likely want to track "files" rather than blocks (since blobs are split into multiple blocks). I can see a way to do this for "download" progress, but tracking uploads to a peer and whether those are "done" is more complicated. Maybe we could listen to and track peer "wants"? Some joint thinking on this could be helpful.

sethvincent commented on tests/helpers/core-manager.js:
```diff
@@ -40,3 +42,37 @@ export function replicate(cm1, cm2) {
   }
 }
+
+export async function waitForCores (coreManager, keys) {
```
I moved these functions into the core-manager helpers file. Haven't imported them into the core-manager tests yet but can do that.

sethvincent commented on tests/helpers/core-manager.js:
```diff
@@ -40,3 +42,37 @@ export function replicate(cm1, cm2) {
+  if (hasKeys(keys, allKeys)) return
+  return new Promise(res => {
+    coreManager.on('add-core', async function onAddCore ({ key, core }) {
+      await core.ready()
```
Added `await core.ready()` here for convenience. Did a quick check to make sure that still works for the core-manager tests and it seems fine.

sethvincent commented on tests/helpers/core-manager.js:
```diff
@@ -7,15 +7,17 @@ import NoiseSecretStream from '@hyperswarm/secret-stream'
     storage: RAM,
-    projectKey
+    projectKey,
+    projectSecretKey,
```
I found it was necessary to be able to pass the secret key here.

sethvincent commented:
I have a few tests to add still but otherwise this is ready for review.

sethvincent commented:
Fixed the weird stalling test :tada:

gmaclennan commented on lib/sync/replication-state.js:
```diff
@@ -0,0 +1,279 @@
+      this.#onPeer(peer)
+    }
+
+    this.#core.on('peer-add', (peer) => {
+      this.#onPeer(peer)
+    })
```
What happens if a peer disconnects before sync is finished (would happen due to wifi being turned off, walking out of range, or the user closing the app)? I think we need a `peer-remove` event handler.

sethvincent replied:
added a `peer-remove` handler and a test for a peer that destroys their replication stream.

gmaclennan commented on lib/sync/replication-state.js:
```diff
@@ -0,0 +1,279 @@
+  async #onPeer(peer) {
+    await this.#core.update()
+
+    this.#core.on('download', (index, byteLength, from) => {
+      if (!from.remotePublicKey.equals(peer.remotePublicKey)) return
+      const state = this.#getLocalState()
+      state.length = getMaxLength(this.#core)
+
+      if (state.have === 0 && state.want === 0) {
+        state.want = state.length
+      }
+
+      state.want--
+      state.have++
+      this.#setPeerState(this.#core.key, state)
+    })
+
```
Need to cleanup this event listener when the peer is disconnected.

gmaclennan reviewed (changes requested):
I haven't read through the tests yet, but reviewed `sync/replication-state.js`. I wouldn't normally comment on performance questions, but I think a lot of this code is in "hot" code paths, that are going to be called multiple times. In general I think it is better for perf to directly use iterators in `for...of` statements, rather than converting them to arrays and then using the array iteration methods (`map`, `reduce`, `filter`). I believe converting to an array is already going to iterate through them once, and then has the memory overhead of creating the array, which then needs to be garbage collected.
Also, while I am a `Array.reduce()` fan, I keep seeing feedback that people find it hard to understand code that uses `reduce()`, so maybe using `for...of` can be better for readability too.

sethvincent commented on lib/sync/replication-state.js:
```diff
@@ -0,0 +1,279 @@
+  }
+
+  close() {
+    this.removeAllListeners()
```
So far this class only adds listeners to itself, consumers don't add listeners. I imagined that this class would be internal to the ReplicationState class that would surface state of all cores to a consumer. If that ends up not being the case then I think this can be revisited.

sethvincent commented on lib/sync/replication-state.js:
```diff
@@ -0,0 +1,279 @@
+   * @param {import('../core-manager/core-index.js').CoreRecord} coreRecord
+   */
+  async #core(coreRecord) {
+    const core = coreRecord.core
```
if the core hasn't reached the ready state the bitfield isn't available. Seems best to make sure here rather than try and expect that the core is already ready to be accessed.

sethvincent commented on lib/sync/replication-state.js:
```diff
@@ -0,0 +1,279 @@
+    let haveLocal = 0
+    let haveRemote = 0
+
+    for (let i = 0; i < length; i++) {
+        const local = localBitfield.get(i)
+        const remote = remoteBitfield.get(i)
+
+        if (local) {
+            haveLocal++
+        }
+
+        if (remote) {
+            haveRemote++
+        }
+    }
```
I added a test for a peer that destroys their stream after downloading only part of peer 1's data (and the opposite). Later the initial two peers have separately synced with a third peer, each getting different parts of peer 3's data, and then the two initial peers re-sync with each other. Does this cover that issue?

The only thing that really needed to be changed was adding the `peer-remove` handler and using `ifAvailable: true` in calls to `core.download()`.

What specifically doesn't work here? Any recommended alternative?

gmaclennan replied:
I've added a failing test that I hope helps explain what I mean. This test _should_ not fail with the core in the hypercore-replication-repo (but I haven't tested it). The difference is that the "original" code counts what is shared, and what is missing from each. This is different that just counting how many blocks each (local, remote) has.

sethvincent commented on lib/sync/replication-state.js:
```diff
@@ -0,0 +1,279 @@
+    const peerOnBitfield = peer.onbitfield
+    const peerOnRange = peer.onrange
+
+    peer.onbitfield = (bitfield) => {
+      peerOnBitfield.call(peer, bitfield)
+      this.#updatePeerState(peer)
+    }
+
+    peer.onrange = ({ drop, start, length }) => {
+      peerOnRange.call(peer, { drop, start, length })
+      this.#updatePeerState(peer)
+    }
```
Yeah I'm not a fan of it either. I'd like to revise `#updatePeerState` to take the drop, start, and length options and use those instead of looping through the whole bitfield.

So far, though, looping through thousands of records doesn't seem to be a significant issue.

I'd lean toward working on this (and a couple of the other perf suggestions) after setting up a way to do benchmarking.

sethvincent commented on lib/sync/replication-state.js:
```diff
@@ -0,0 +1,279 @@
+      this.#onPeer(peer)
+    }
+
+    this.#core.on('peer-add', (peer) => {
```
I _think_ it'll have the same lifecycle as the app.

Oh, or in multi-projects land it might be equivalent to the lifecycle of the active project.

Unless we have a sync party mode that syncs multiple projects at the same time.

I thought I had an answer to this and then quickly realized I didn't, haha.

sethvincent commented:
@gmaclennan I addressed almost all of the comments. I'm not yet certain what to do with removing some of the hypercore event handlers.

I fixed some of the perf suggestions and made an issue for a few others: #99 
And an issue for setting up some benchmarking: #100 

I think it would be nice to work on the perf stuff after setting up benchmarking so it can be tracked.

gmaclennan reviewed (changes requested):
I added another failing test for syncing 3 peers when one has blocks unavailable.

The issue is the way that "local state" is used: the `coreReplicationState.#updatePeerState()` method updates the local state for the core, but what is being written is the local state in relation to that particular peer (e.g. what data does that peer have that I need). This "local state" is different for each peer. This means that the local state is overwritten with every peer update, and it is unclear what the `'synced'` event means. I think with the current code it means that the core has synced everything it can with a particular peer, but the code seems to read that as the core having finished syncing with _all_ peers.

The other thing to consider here is how the overall replication state is useful for the frontend / user. As I understand the core, the "replication state" is a map of core id to the state of sync with _all peers_ (although see comment above, I think it is actually just the most recently updated peer). Based on the designs we have for showing sync state, what is actually useful for the user is the sync state with each peer, across all cores (e.g. the breakdown by core is more an implementation detail and not so useful for the frontend). What the user needs to know is "have I synced everything I need to with a particular peer?", or "how much data do I need to sync with a particular peer?".

This is all very complicated and I find it hard to get my head around, but I think what we want is:

For each peer: how much data do they already have from us? and how much data do I have available that they still want? The latter is how we know whether sync with a particular peer is "complete", and the former is how we can show an idea of percentage "progress"

For the local peer, across all peers, what data is available that I already have? And what data is available _across all peers_ that I want?


Sorry to keep coming back with more difficult questions here, but I think this part of the code is really important to get right, and the logic is complicated but errors will make the UX confusing for the users (this is one of the main issues we have with sync right now).

sethvincent commented:
@gmaclennan This pr is likely ready for merge now unless there's anything else to fix first. We could probably come up with more tests in follow-up PRs.

gmaclennan commented:
Hi Seth, I finally had a chance to review this again. Sorry about the delay.

I still see issues with this PR with race conditions related to local state which I described above:

> The issue is the way that "local state" is used: the coreReplicationState.#updatePeerState() method updates the local state for the core, but what is being written is the local state in relation to that particular peer (e.g. what data does that peer have that I need). This "local state" is different for each peer. This means that the local state is overwritten with every peer update, and it is unclear what the 'synced' event means. I think with the current code it means that the core has synced everything it can with a particular peer, but the code seems to read that as the core having finished syncing with all peers.

When I console.log state for the 'sync' event for the 'replicate 3 cores with unavailable blocks' test I see that sometimes the 'sync' event is firing 3 times, and sometimes once. When the `sync` event does only fire once, the final state for the local core (`core3`) is incorrect:

```
{"coreId":"78acdcf","remote":false,"have":7,"want":0,"unavailable":3,"length":7}
```

However when the `sync` event fires 3 times (I just keep running to test to see this) then the final state (on the final `sync` event) has the correct state for local:

```
{"coreId":"45d7bb7","remote":false,"have":7,"want":0,"unavailable":0,"length":7}
```

I think this issue is important to resolve because it will result in incorrect information being reported about whether sync is complete or not, which has been a big issue for users who need to know when they can consider sync as done.

I believe this issue boils down to the local state that we are reading from the bitfields specific to the pair of cores that are replicating, which is different depending on which pair is replicating, e.g.

Core A (local) is replicating with Core B, where Core B is missing 3 records. Core A state here will have `unavailable: 3`.
Core A (local) is replicating with Core C, where Core C has all records. Core A state here will have `unavailable: 0`.

Once Core A has downloaded all blocks (which is can do, because it downloads them from Core C), then it should read `unavailable: 0`, but if the latest state update comes from the replication with Core B, then the "local state" of that replication overwrites the "overall" local state, resulting in an incorrect state.

I'm happy to get on a call to discuss this and review the best solution if that would be helpful.

sethvincent commented:
Ah, right, I'm seeing that now. I'll work on revising the way local state is managed.

sethvincent commented:
I found a way to resolve the incorrect state in that last test. With the exception of tracking unavailable blocks all the information about local core state comes from the local core. Previously it was doing what you describe (using peer info that overwrites local state info) with just the unavailable block count.

I think the `CoreReplicationState` class is now pretty solid. I'm not so sure about the higher-level `ReplicationState` class. It works but I need to use it in a more real-world situation to get a better sense of the events being emitted and the shape the emitted data ought to be.
